test/slipstream/socket_test_test.exs

```diff
@@ -0,0 +1,193 @@
+  clients: in order to make assertions about the status of the client, we have
+  written a client fixture (`@fixture`) that sends messages to the test process
+  when meaningful state changes occur (on connection, e.g.). This is not
+  recommended for writing clients in the wild.
+  """
+  @moduledoc since: "0.2.0"
+
+  @fixture Slipstream.TestModeClient
+
+  setup do
+    [topic: "rooms:lobby", event: "msg:new", params: %{"fizz" => "buzz"}]
+  end
+
+  describe "given that the fixture is spawned" do
+    setup do
+      start_supervised!({@fixture, self()})
+
+      :ok
+    end
+
+    test "the client connects when we do an accept_connect/1" do
+      accept_connect(@fixture)
+
+      assert_receive {@fixture, :connected}
+    end
+
+    test "passing a nil client or fake name raises ArgumentError" do
+      assert_raise ArgumentError, ~r"cannot find pid for client nil", fn ->
+        accept_connect(nil)
+      end
+
+      assert_raise ArgumentError, ~r"cannot find pid for client :foo", fn ->
+        accept_connect(:foo)
+      end
+    end
+
+    test "passing a dead client will raise ArgumentError" do
+      pid = spawn(fn -> exit(:normal) end)
+      ref = Process.monitor(pid)
+
+      assert_receive {:DOWN, ^ref, :process, ^pid, :normal}
+      refute Process.alive?(pid)
+
+      assert_raise ArgumentError,
+                   ~r"cannot send to client #PID<[\d\.]+> that is not alive",
+                   fn ->
+                     accept_connect(pid)
+                   end
+    end
+
+    test "we may connect and join a topic all at once with connect_and_join/5",
+         c do
+      topic = c.topic
+      params = c.params
+
+      join(topic, params)
+
+      connect_and_join @fixture, ^topic, ^params, {:ok, params}
+
+      assert_receive {@fixture, :connected}
+      assert_receive {@fixture, :joined, ^topic, ^params}
+    end
+
+    test "the error atom returns a join-crashed response", c do
+      topic = c.topic
+      params = c.params
+
+      join(topic, params)
+
+      connect_and_join @fixture, ^topic, ^params, :error
+
+      assert_receive {@fixture, :connected}
+
+      assert_receive {@fixture, :topic_closed, ^topic,
+                      {:failed_to_join, %{"error" => "join crashed"}}}
+    end
+  end
+
+  describe "given the client is connected" do
+    setup do
+      start_supervised!({@fixture, self()})
+
+      accept_connect(@fixture)
+
+      assert_receive {@fixture, :connected}
+
+      :ok
+    end
+
+    test "disconnect/2 disconnects the client" do
+      reason = :bye!
+
+      disconnect(@fixture, reason)
+
+      assert_receive {@fixture, :disconnected, ^reason}
+    end
+
+    test "assert_join/4 allows a topic join", c do
+      topic = c.topic
+      params = c.params
+
+      join(topic, params)
+
+      assert_join ^topic, ^params, :ok
+
+      assert_receive {@fixture, :joined, ^topic, %{}}
+
+      # once joined, we do not duplicate the join
+      refute_join ^topic, ^params
+    end
```

Comment:
I'm guessing this is the mindfuck of testing a test framework, but what's going on here? How does `assert_join` "allow" a `:joined` message?

Reply:
I guess I see the `assert_join` macro doing its own `assert_receive` on the join command. ok, ofc, SocketTest is mocking the other end of the socket and sending back a `:joined` event.

so, breaking down the 3 key lines...

```
# I'm the client. join this topic
join(topic, params)

# assert that the socket process sees the join command from above
assert_join ^topic, ^params, :ok

# assert that this test process (the client) sees its :joined confirmation
assert_receive {@fixture, :joined, ^topic, %{}}
```

tell me where I'm being imprecise, or dead wrong...

Reply:
I think I'm gonna rename the `defp join/2` to `defp tell_fixture_to_join/2` because it looks like something that would be provided by Slipstream.SocketTest but it's not

but yep you have the right idea. Check out the fixture in `test/fixtures/test_mode_client.ex`, specifically its implementation of `c:Slipstream.handle_join/3`. It does a `Kernel.send(socket.assigns.test_proc, {__MODULE__, :joined, topic, params})`, which we can assert happened with that `assert_receive/2`

to break it down like you:

```elixir
# tell the client to join the server's "foo" topic
join("foo", params)

# assert that the client is trying to join "foo" and let it join (the `:ok` part)
assert_join "foo", ^params, :ok

# assert that the client was notified that it has joined succesfully
# and that its `c:Slipstream.handle_join/3` callback was invoked
assert_receive {@fixture, :joined, ^topic, %{}}
```

Reply:
renamed that helper to `tell_fixture_to_join/2` and referenced it everywhere with `__MODULE__.tell_fixture_to_join/2` just to make it super clear that it's not imported